File: skills/skill-creator/scripts/runners/goose.py
#!/usr/bin/env python3
"""Goose runner for skill description and trigger evaluation."""

from __future__ import annotations

import json
import os
import shlex
import subprocess
from pathlib import Path

from .base import Runner, RunnerError


class GooseRunner(Runner):
    def __init__(self, command: str | None = None):
        raw = command or os.environ.get("SKILL_CREATOR_GOOSE_COMMAND", "goose")
        self.command = tuple(shlex.split(raw))
        if not self.command:
            raise RunnerError("SKILL_CREATOR_GOOSE_COMMAND cannot be empty")

    def _text_command(self, model: str | None) -> list[str]:
        command = [
            *self.command,
            "run",
            "--no-session",
            "--quiet",
            "--output-format",
            "text",
            "--instructions",
            "-",
        ]
        if model:
            command.extend(["--model", model])
        return command

    def _stream_command(self, query: str, model: str | None) -> list[str]:
        command = [
            *self.command,
            "run",
            "--no-session",
            "--quiet",
            "--output-format",
            "stream-json",
            "--text",
            query,
        ]
        if model:
            command.extend(["--model", model])
        return command

    def run_text(
        self,
        prompt: str,
        model: str | None,
        timeout: int = 300,
        cwd: Path | None = None,
    ) -> str:
        command = self._text_command(model)
        try:
            result = subprocess.run(
                command,
                input=prompt,
                capture_output=True,
                text=True,
                cwd=cwd,
                timeout=timeout,
            )
        except FileNotFoundError as error:
            raise RunnerError(f"Goose command not found: {self.command[0]}") from error
        if result.returncode != 0:
            raise RunnerError(
                f"Goose exited {result.returncode}: {' '.join(command)}\n"
                f"stderr: {result.stderr.strip()}"
            )
        return result.stdout

    def start_stream(
        self,
        query: str,
        model: str | None,
        cwd: Path,
    ) -> subprocess.Popen:
        command = self._stream_command(query, model)
        try:
            return subprocess.Popen(
                command,
                stdout=subprocess.PIPE,
                stderr=subprocess.DEVNULL,
                cwd=cwd,
            )
        except FileNotFoundError as error:
            raise RunnerError(f"Goose command not found: {self.command[0]}") from error

    def event_loaded_skill(self, event: dict, skill_name: str) -> bool | None:
        event_type = event.get("type")
        if event_type in {"complete", "error"}:
            return False
        if event_type != "message":
            return None

        message = event.get("message", {})
        for content in message.get("content", []):
            if content.get("type") not in {"toolRequest", "tool_request"}:
                continue
            tool_call = content.get("toolCall", content.get("tool_call", {}))
            if isinstance(tool_call, dict) and "Ok" in tool_call:
                tool_call = tool_call["Ok"]
            if isinstance(tool_call, dict) and "value" in tool_call:
                tool_call = tool_call["value"]
            if not isinstance(tool_call, dict):
                continue
            name = str(tool_call.get("name", ""))
            arguments = tool_call.get("arguments", {})
            if name.rsplit("__", 1)[-1] != "load_skill":
                continue
            if isinstance(arguments, str):
                try:
                    arguments = json.loads(arguments)
                except json.JSONDecodeError:
                    arguments = {}
            loaded = arguments.get("name", "") if isinstance(arguments, dict) else ""
            return loaded == skill_name or loaded.startswith(f"{skill_name}/")
        return None

File: skills/skill-creator/scripts/runners/goose.ts
// Goose runner for skill description and trigger evaluation.
import { spawn } from "node:child_process";
import { RunnerError } from "./base.js";
import type { Runner, StreamProcess } from "./base.js";

function shlexSplit(command: string): string[] {
  const parts: string[] = [];
  const regex = /"([^"]*)"|'([^']*)'|(\S+)/g;
  let match: RegExpExecArray | null;
  while ((match = regex.exec(command)) !== null) {
    parts.push(match[1] ?? match[2] ?? match[3]);
  }
  return parts;
}

export class GooseRunner implements Runner {
  command: string[];

  constructor(command?: string) {
    const raw = command || process.env.SKILL_CREATOR_GOOSE_COMMAND || "goose";
    this.command = shlexSplit(raw);
    if (!this.command.length) {
      throw new RunnerError("SKILL_CREATOR_GOOSE_COMMAND cannot be empty");
    }
  }

  textCommand(model: string | null): string[] {
    const command = [...this.command, "run", "--no-session", "--quiet", "--output-format", "text", "--instructions", "-"];
    if (model) command.push("--model", model);
    return command;
  }

  streamCommand(query: string, model: string | null): string[] {
    const command = [...this.command, "run", "--no-session", "--quiet", "--output-format", "stream-json", "--text", query];
    if (model) command.push("--model", model);
    return command;
  }

  async runText(prompt: string, model: string | null, timeout = 300, cwd?: string): Promise<string> {
    const command = this.textCommand(model);
    const [bin, ...args] = command;

    return new Promise((resolvePromise, reject) => {
      const child = spawn(bin, args, { cwd, timeout: timeout * 1000 });
      let stdout = "";
      let stderr = "";
      child.stdout.on("data", (chunk) => (stdout += chunk));
      child.stderr.on("data", (chunk) => (stderr += chunk));
      child.on("error", (error: NodeJS.ErrnoException) => {
        if (error.code === "ENOENT") {
          reject(new RunnerError(`Goose command not found: ${this.command[0]}`));
        } else {
          reject(error);
        }
      });
      child.on("close", (code) => {
        if (code !== 0) {
          reject(
            new RunnerError(
              `Goose exited ${code}: ${command.join(" ")}\nstderr: ${stderr.trim()}`
            )
          );
          return;
        }
        resolvePromise(stdout);
      });
      child.stdin.write(prompt);
      child.stdin.end();
    });
  }

  startStream(query: string, model: string | null, cwd: string): StreamProcess {
    const command = this.streamCommand(query, model);
    const [bin, ...args] = command;
    try {
      return spawn(bin, args, { cwd, stdio: ["ignore", "pipe", "ignore"] });
    } catch (error) {
      throw new RunnerError(`Goose command not found: ${this.command[0]}`);
    }
  }

  eventLoadedSkill(event: Record<string, unknown>, skillName: string): boolean | null {
    const eventType = event.type;
    if (eventType === "complete" || eventType === "error") return false;
    if (eventType !== "message") return null;

    const message = (event.message as Record<string, unknown>) ?? {};
    const content = (message.content as Array<Record<string, unknown>>) ?? [];
    for (const item of content) {
      if (!["toolRequest", "tool_request"].includes(item.type as string)) continue;
      let toolCall = (item.toolCall ?? item.tool_call ?? {}) as Record<string, unknown>;
      if (typeof toolCall === "object" && toolCall !== null && "Ok" in toolCall) {
        toolCall = toolCall.Ok as Record<string, unknown>;
      }
      if (typeof toolCall === "object" && toolCall !== null && "value" in toolCall) {
        toolCall = toolCall.value as Record<string, unknown>;
      }
      if (typeof toolCall !== "object" || toolCall === null) continue;
      const name = String(toolCall.name ?? "");
      let args = toolCall.arguments ?? {};
      if (name.split("__").pop() !== "load_skill") continue;
      if (typeof args === "string") {
        try {
          args = JSON.parse(args);
        } catch {
          args = {};
        }
      }
      const loaded = typeof args === "object" && args !== null ? String((args as any).name ?? "") : "";
      return loaded === skillName || loaded.startsWith(`${skillName}/`);
    }
    return null;
  }
}
